In [1]:
import pandas as pd

In [4]:
data = pd.read_csv('/home/mayowafunmi/data_science_tutorial/pythonfiles/police_data.csv')
data

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

In [5]:
data.count()

stop_date             65535
stop_time             65535
country_name              0
driver_gender         61474
driver_age_raw        61481
driver_age            61228
driver_race           61475
violation_raw         61475
violation             61475
search_conducted      65535
search_type            2479
stop_outcome          61475
is_arrested           61475
stop_duration         61475
drugs_related_stop    65535
dtype: int64

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

In [10]:
# display number of values that has no record
data.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

In [15]:
# remove columns that only contain null values
#data.drop(columns='country_name', inplace=True)

In [16]:
data.head(2)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False

In [19]:
# for speeding, were men or women stopped more often?
# check the violation column for Speeding
data[data['violation']=='Speeding']

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
4      3/14/2005     10:00             F          1984.0        21.0   
6       4/1/2005     17:30             M          1969.0        36.0   
...          ...       ...           ...             ...         ...   
65527  12/6/2012     15:26             F          1981.0        31.0   
65529  12/6/2012     16:00             M          1994.0        18.0   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race violation_raw violation  search_conducted search_type  \
0           White      Speeding  Speeding             False         NaN   
1           White      Speeding  Speeding             False         NaN   
2           White      Speeding  Speeding             False         NaN   
4           White      Speeding  Speeding             False         NaN   
6           White      Speeding  Speeding             False         NaN   
...           ...           ...       ...               ...         ...   
65527       White      Speeding  Speeding             False         NaN   
65529       White      Speeding  Speeding             False         NaN   
65530       White      Speeding  Speeding             False         NaN   
65531       White      Speeding  Speeding             False         NaN   
65534       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
0         Citation       False      0-15 Min               False  
1         Citation       False      0-15 Min               False  
2         Citation       False      0-15 Min               False  
4         Citation       False      0-15 Min               False  
6         Citation       False      0-15 Min               False  
...            ...         ...           ...                 ...  
65527     Citation       False      0-15 Min               False  
65529     Citation       False      0-15 Min               False  
65530     Citation       False      0-15 Min               False  
65531      Warning       False      0-15 Min               False  
65534     Citation       False      0-15 Min               False  

[37204 rows x 14 columns]

In [20]:
# count the number of driver_gender for the speeding violation column
data[data['violation']=='Speeding'].driver_gender.value_counts()

M    25517
F    11686
Name: driver_gender, dtype: int64

In [23]:
# does gender affect who gets searched during a stop
# consider two columns: driver_gender and search_conducted

In [26]:
data

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 14 columns]

In [28]:
# get all F in driver_gender
data.groupby('driver_gender').get_group('F')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
4      3/14/2005     10:00             F          1984.0        21.0   
7       6/6/2005     13:20             F          1986.0        19.0   
11     7/13/2005     19:00             F          1966.0        39.0   
13     7/18/2005     19:30             F          1984.0        21.0   
19     7/24/2005     20:10             F          1958.0        47.0   
...          ...       ...           ...             ...         ...   
65521  12/6/2012     14:00             F          1992.0        20.0   
65524  12/6/2012     15:06             F          1974.0        38.0   
65527  12/6/2012     15:26             F          1981.0        31.0   
65530  12/6/2012     17:54             F          1987.0        25.0   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race violation_raw violation  search_conducted search_type  \
4           White      Speeding  Speeding             False         NaN   
7           White      Speeding  Speeding             False         NaN   
11          White      Speeding  Speeding             False         NaN   
13          White      Speeding  Speeding             False         NaN   
19          White      Speeding  Speeding             False         NaN   
...           ...           ...       ...               ...         ...   
65521       White      Speeding  Speeding             False         NaN   
65524       White      Speeding  Speeding             False         NaN   
65527       White      Speeding  Speeding             False         NaN   
65530       White      Speeding  Speeding             False         NaN   
65534       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
4         Citation       False      0-15 Min               False  
7         Citation       False      0-15 Min               False  
11        Citation       False      0-15 Min               False  
13        Citation       False      0-15 Min               False  
19        Citation       False      0-15 Min               False  
...            ...         ...           ...                 ...  
65521     Citation       False      0-15 Min               False  
65524     Citation       False      0-15 Min               False  
65527     Citation       False      0-15 Min               False  
65530     Citation       False      0-15 Min               False  
65534     Citation       False      0-15 Min               False  

[16310 rows x 14 columns]

In [29]:
data.groupby('driver_gender').search_conducted.sum()

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

In [31]:
data.search_conducted.value_counts()

False    63056
True      2479
Name: search_conducted, dtype: int64

# This shows that search was conducted 2479 times, and out of this, female were searched 366 times while male 2113 times

# 

# What is the mean stop_duration?

In [52]:
data.head(2)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   

  is_arrested  stop_duration  drugs_related_stop  
0       False            NaN               False  
1       False            NaN               False

In [34]:
# convert stop_duration datatype to integers
# first, get all the values in the column
data.stop_duration.value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: stop_duration, dtype: int64

In [35]:
data.stop_duration.unique()

array(['0-15 Min', '16-30 Min', '30+ Min', nan, '2'], dtype=object)

In [36]:
data.stop_duration.nunique()

4

In [53]:
# change the values to the average of the values
data['stop_duration'] = data['stop_duration'].map({'0-15 Min': 7.5, '16-30 Min': 23, '30+ Min': 45})

In [63]:
data = pd.read_csv('/home/mayowafunmi/data_science_tutorial/pythonfiles/police_data.csv')
data

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

In [64]:
data['stop_duration'] = data['stop_duration'].map({'0-15 Min': 7.5, '16-30 Min': 23, '30+ Min': 45})

In [65]:
data

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested  stop_duration  \
0                 False         NaN       Citation       False            7.5   
1                 False         NaN       Citation       False            7.5   
2                 False         NaN       Citation       False            7.5   
3                 False         NaN  Arrest Driver        True           23.0   
4                 False         NaN       Citation       False            7.5   
...                 ...         ...            ...         ...            ...   
65530             False         NaN       Citation       False            7.5   
65531             False         NaN        Warning       False            7.5   
65532             False         NaN       Citation       False            7.5   
65533             False         NaN            NaN         NaN            NaN   
65534             False         NaN       Citation       False            7.5   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

In [66]:
# the mean

In [67]:
data['stop_duration'].mean()

12.001195627419722

 # compare the age distributions for each violation

In [68]:
data.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  search_type stop_outcome is_arrested  stop_duration  drugs_related_stop  
0         NaN     Citation       False            7.5               False  
1         NaN     Citation       False            7.5               False

In [70]:
data.groupby('violation').driver_age.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0

In [71]:
data.groupby('driver_age').violation.describe()

count unique               top  freq
driver_age                                     
15.0           5      2  Moving violation     4
16.0          34      5          Speeding    18
17.0         449      5          Speeding   338
18.0        1344      5          Speeding   980
19.0        2388      5          Speeding  1655
...          ...    ...               ...   ...
83.0           2      2          Speeding     1
84.0           3      1          Speeding     3
85.0           1      1  Moving violation     1
86.0           6      3          Speeding     3
88.0           2      1          Speeding     2

[73 rows x 4 columns]